File: archive/proj8/proj_8_testers/tests/test7.ml
# Pass by reference test

begin
    foo3(int ref x) begin
        int y;
        write("foo3: ", x, "\n");
		read(y);
		x := x + y;
		write("foo3: ", x, "\n");
    end

    foo2(int ref x) begin
        int z;
        read(z);
        x := x + z;
        write("foo2: ", x, "\n");
        foo3(x)
        write("foo2: ", x, "\n");
    end

    foo1(int ref x) begin
        int t;
        read(t);
        write("foo1: ", x, "\n");
        x := t * x;
        write("foo1: ", x, "\n");
    end

    int x;
    read(x);
    write("main: ", x, "\n");
    foo1(x)
    write("main: ", x, "\n");
    foo2(x)
    write("main: ", x, "\n");
end

File: archive/proj8/proj_8_testers/c_files/test7.c
// Pass by reference
// You are not obligated to use this syntax.  
// Pascal uses the keyword "var": "integer foo(var x: integer)".
// C++ uses the & symbol following the parameter name.  
// Using the key word "ref" is another possibility.
#include <stdio.h>

void foo3(int* x) {
  int y;
  printf("foo3: %d\n", *x);
  scanf("%d", &y);
  *x = *x + y;
  printf("foo3: %d\n", *x);
}

void foo2(int* x) {
  int z;
  scanf("%d", &z);
  *x = *x + z;
  printf("foo2: %d\n", *x);
  foo3(x);
  printf("foo2: %d\n", *x);
}

void foo1(int* x) {
  int t;
  scanf("%d", &t);
  printf("foo1: %d\n", *x);
  *x = t*(*x);
  printf("foo1: %d\n", *x);
}

int main() {
  int x;
  scanf("%d", &x);
  printf("main: %d\n", x);
  foo1(&x);
  printf("main: %d\n", x);

  foo2(&x);
  printf("main: %d\n", x);
  return 0;
}
